### PR TITLE
[tx] implement model unloading

### PR DESCRIPTION
**Pull request by raulchen**
Closes https://github.com/NovaSky-AI/SkyRL/issues/841

This PR implements model unloading (manual via `unload_model` or automatic on session timeout). 

# Changes

New API endpoint:
- POST /api/v1/unload_model - unloads a model and frees its LoRA adapter slot

Automatic garbage collection:
- Engine periodically checks for sessions with stale heartbeats (configurable via --session-cleanup-interval-sec and --sessio
n-timeout-sec)
- Models from expired sessions are automatically unloaded, freeing adapter slots for reuse

Backend changes:
- AbstractBackend.delete_model() - new abstract method
- JaxBackend.delete_model() - clears LoRA weights and frees adapter index
- clear_adapter_config() - zeros out lora_A, lora_B, scaling, and rank for an adapter
- Adapter index allocation now reuses freed slots instead of always incrementing

# Test plan

- Integration tests in `test_api.py`
- unit tests in "test_engine.py" and "test_jax_backend.py"

### CODE REVIEW

**Comment by pcmoritz** [comment]
Thanks a lot for implementing this! Some suggestions before merging this:

- Remove the manual unload API call -- I think it is too early to think about diverging from the upstream API / SDK (in the best case it is an extra API call that people don't really use and in the worst case there will be similar functionality upstream at some point with a [slightly] different API and then we need to reconcile it -- better to not add it until we think it is really needed)
- Make it possible to turn off this functionality by setting `session_cleanup_interval_sec` and/or `session_timeout_sec` to `-1` respectively (I'd suggest doing this since it is a very simple change, technically it can already be achieved by setting the timeouts to very large values, but it is probably better to have a documented way to do this)
- Add and index for ModelDB.status, since it will make sure the queries in cleanup_stale_sessions can be done in constant time

**Comment by raulchen** [comment]
The manual `unload_model` API [already exists](https://github.com/thinking-machines-lab/tinker/blob/2d8e9d5e00f746f39148a5d0cb760dff3f2eed43/src/tinker/resources/models.py#L153) in Tinker

Will address the other 2 comments.

**Comment by pcmoritz** [comment]
Got it that makes sense, can you try to modify the test code to use the SDK calls instead of plain http?

File: skyrl-tx/tests/tinker/test_jax_backend.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
+"""Unit tests for JaxBackend."""
```
Comment (raulchen) [inline skyrl-tx/tests/tinker/test_jax_backend.py:1]: Current test_engine.py mixes tests for both TinkerEngine and JaxBackend. 
I'd be nice to separate them as the implementations have been separated. 
I've done so for the new tests in this PR. 
Will refactor the remaining tests in a follow-up PR.

File: skyrl-tx/tx/layers/lora.py [hunk 2]
```diff
@@ -70,7 +70,7 @@ def apply_lora(
 
         (batch_size, seq_len, *dims) = x.shape
         assert len(self.lora_A.shape) == 3
-        assert len(dims) == 0 if isinstance(self, nnx.Embed) else tuple(dims) == self.lora_A.value.shape[1:-1]
+        assert len(dims) == 0 if isinstance(self, nnx.Embed) else tuple(dims) == self.lora_A[...].shape[1:-1]
```
Comment (raulchen) [inline skyrl-tx/tx/layers/lora.py:73]: this is a drive-by fix because ".value" is deprecated.

File: skyrl-tx/tx/tinker/backends/jax.py [hunk 3]
```diff
@@ -416,11 +416,12 @@ def create_model(self, model_id: str, lora_config: types.LoraConfig) -> None:
...
-        # Allocate adapter index for this model_id
-        adapter_index = max((m.adapter_index for m in self.models.values()), default=0) + 1
-
-        if adapter_index >= self.config.max_lora_adapters:
+        # Allocate adapter index for this model_id (find first available slot)
+        used_indices = {m.adapter_index for m in self.models.values()}
+        available_indices = set(range(1, self.config.max_lora_adapters + 1)) - used_indices
+        if not available_indices:
             raise ValueError(f"Maximum number of LoRA adapters ({self.config.max_lora_adapters}) reached")
+        adapter_index = min(available_indices)
```
[the remark this answers is not in the document]
Comment (raulchen) [reply skyrl-tx/tx/tinker/backends/jax.py:425]: nice catch. 
so there are actually "max_lora_adapters - 1" available slots for the user. 
This might be confusing and unintuitive to users. should we  change the behavior? @pcmoritz

File: skyrl-tx/tx/tinker/engine.py [hunk 4]
```diff
@@ -344,6 +347,79 @@ def process_create_model(self, model_id: str, request_data: types.CreateModelInp
...
         )
 
+    def process_unload_model(self, model_id: str, request_data: types.UnloadModelInput) -> types.UnloadModelOutput:
+        """Unload a model and free all resources."""
+        if self.backend.has_model(model_id):
+            self.backend.delete_model(model_id)
+        else:
+            logger.warning(f"Ignoring unload request for model {model_id} that is not loaded.")
+
+        self.backend.delete_model(model_id)
```
Comment (pcmoritz) [inline skyrl-tx/tx/tinker/engine.py]: This was already called above, I think it is a mistake to call it again, right?

Comment (pcmoritz) [reply skyrl-tx/tx/tinker/engine.py]: (actually this is probably the reason why the test is currently failing)

File: skyrl-tx/tx/tinker/types.py [hunk 5]
```diff
@@ -80,6 +81,15 @@ class CreateModelOutput(BaseModel):
     lora_config: LoraConfig
 
 
+class UnloadModelInput(BaseModel):
+    pass
+
+
+class UnloadModelOutput(BaseModel):
+    model_id: str
+    status: str = "unloaded"
```
Comment (pcmoritz) [inline skyrl-tx/tx/tinker/types.py]: Let's remove the default for this, I tried to keep defaults out of types in `types.py` (haven't always succeeded, but if we can remove more, that's a win I think). Defaults can lead to bugs where we forget to fill the value and it will be the wrong value because of that :)

File: skyrl-tx/tx/layers/lora.py (hunk 2 above)
Comment (pcmoritz) [reply skyrl-tx/tx/layers/lora.py:73]: very nice, thanks for fixing this!

File: skyrl-tx/tests/tinker/test_jax_backend.py (hunk 1 above)
Comment (pcmoritz) [reply skyrl-tx/tests/tinker/test_jax_backend.py:1]: Thanks for getting this started, there is also an issue about it https://github.com/NovaSky-AI/SkyRL/issues/819 :)

File: skyrl-tx/tx/tinker/engine.py (hunk 4 above)
Comment (raulchen) [reply skyrl-tx/tx/tinker/engine.py]: it's an accidental bug introduced in the last commit. fixed now

**Comment by OhadRubin** [comment]
I should probably open an issue for this but:
@raulchen @pcmoritz clear_adapter_config results in no gradient flow on subsequent non initial runs since both A and B are 0.
Should init the cleared loras weights with with the original default initilizers.
Or Al least only zero out one of the matrices (I think A is initted with zeros usually)

**Comment by pcmoritz** [comment]
@OhadRubin Yes you are right! I briefly thought about this when reviewing the PR and thought it was reinitializing when the new adapter is allocated but that's not true. The best solution is probable to fix that. Let me know if you would like to make a PR for that, otherwise I'm happy to do it too :)

**Comment by raulchen** [comment]
Thanks for reporting this issue. I can also put up a fix later today in case you don't have bandwidth.

**Comment by pcmoritz** [comment]
@raulchen Why don't you make a PR for it (in `update_adapter_config`, get the initializer of `lora_A` and reinitialize it, and we should do that before we clear the out-of-rank entries) and we can discuss more in the PR :)

I propose we also rename `update_adapter_config` to `reinitialize_adapter` so it is clear what this function does. And we should also initialize lora_B to zero so everything is clear and ready and we don't depend on the previous values.

I think it is best to do it when creating the model instead of doing it when deleting the model -- when the model is deleted, the responsibility should be to ensure no information from this user leaks, when the model is initialized, the responsibility should be to make sure it is ready for the next training.

**Comment by raulchen** [comment]
hey @OhadRubin @pcmoritz , here is the fix https://github.com/NovaSky-AI/SkyRL/pull/852
